src/mscorlib/src/System/String.Comparison.cs

```diff
@@ -134,6 +134,48 @@ private unsafe static bool EqualsHelper(String strA, String strB)
             }
         }
 
+        private unsafe static bool EqualsIgnoreCaseAsciiHelper(String strA, String strB)
+        {
+            Contract.Requires(strA != null);
+            Contract.Requires(strB != null);
+            Contract.Requires(strA.Length == strB.Length);
+            Contract.EndContractBlock();
+            int length = strA.Length;
+
+            fixed (char* ap = &strA.m_firstChar) fixed (char* bp = &strB.m_firstChar)
+            {
+                char* a = ap;
+                char* b = bp;
+
+                while (length != 0)
+                {
+                    int charA = *a;
+                    int charB = *b;
+
+                    Debug.Assert((charA | charB) <= 0x7F, "strings have to be ASCII");
+
+                    // Ordinal equals or lowercase equals if the result ends up in the a-z range 
+                    if (charA == charB ||
+                       ((charA | 0x20) == (charB | 0x20) &&
+                          (uint)((charA | 0x20) - 'a') <= (uint)('z' - 'a')))
+                    {
+                        a++;
+                        b++;
+                        length--;
+                    }
+                    else
+                    {
+                        goto ReturnFalse;
```

Comment:
Unfortunately `goto` is the fastest way to return from a loop. About 30% in this particular case compared to a `return`. `break` is good too but requires additional code to be added after the loop to determine how to return.
I'm just guessing here but it might be due to the CPUs complexity rules for the loop stream detector. The compiled code is fine, might be one for the JIT.